### PR TITLE
Migrate `bevy_transform` to required components

### PR DESCRIPTION
**Pull request by ecoskey**
The first step in the migration to required components! This PR removes `GlobalTransform` from all user-facing code, since it's now added automatically wherever `Transform` is used.

## Testing

- None of the examples I tested were broken, and I assume breaking transforms in any way would be visible *everywhere*

---

## Changelog

- Make `Transform` require `GlobalTransform`
~~- Remove `GlobalTransform` from all engine bundles~~
- Remove in-engine insertions of GlobalTransform and TransformBundle
- Deprecate `TransformBundle`
- update docs to reflect changes

## Migration Guide

Replace all insertions of `GlobalTransform` and/or `TransformBundle` with `Transform` alone.

### CODE REVIEW

**Comment by JMS55** [comment]
Missed MaterialMeshletMeshBundle

**Comment by ecoskey** [comment]
oops

File: crates/bevy_transform/src/bundles.rs
Comment (alice-i-cecile) [inline crates/bevy_transform/src/bundles.rs:32]:
```suggestion
#[deprecated(since = "0.15.0", note = "Use `Transform` on its own: `GlobalTransform` is now a required component.")]
```

**Comment by alice-i-cecile** [comment]
I'm in favor of this. The opposite direction (GlobalTransform -> Transform) is much more intrusive and interferes with our plans to split apart Transform (#1275), but I think that this is nearly universally desirable.

**Review by TrialDragon — CHANGES_REQUESTED** [review]
The file `/errors/B0004.md` contains rust doc which uses Transform bundles and should be updated to use this change. (It's why CI is failing). Otherwise, looks good.

**Comment by ecoskey** [comment]
> The file `/errors/B0004.md` contains rust doc which uses Transform bundles and should be updated to use this change. (It's why CI is failing). Otherwise, looks good.

It seems this error would be obsolete or inconsistent with the required components workflow. Should we block this PR on updating things like this or leave that to a later documentation pass?

**Comment by TrialDragon** [comment]
> It seems [the guidance to fix] this error would be obsolete or inconsistent with the required components workflow. Should we block this PR on updating things like this or leave that to a later documentation pass?

I'm not entirely sure it would be? I don't think required components add components to the parent entities, do they? Either way, this can be a follow-up if it is actually redundant in a post required components world.

(Also, the bundle docs seem to be having an issue with `GlobalTransform` no longer being in scope).

**Comment by ecoskey** [comment]
I don't think I expressed that clearly lol. What I meant was that we're going to need to update a lot of docs anyway for the new workflow, so I was wondering if it might make more sense to do a later pass rather than make a lot of changes to that page just because it referenced `TransformBundle`. But doctests need to pass, so I can fix it in the morning :)

**Comment by ecoskey** [comment]
Fixed the docs. They're (specifically errors/B0004.md) in a bit of a weird half-way state since we're intending to nuke SpatialBundle but haven't yet.

File: crates/bevy_transform/src/bundles.rs
Comment (tim-blackbird) [inline crates/bevy_transform/src/bundles.rs:31]: We shouldn't change this struct. We want to show the deprecation notice, not cause unnecessary compilation errors

In my opinion we can leave the docs as they were as well

File: crates/bevy_transform/src/bundles.rs
Comment (tim-blackbird) [inline crates/bevy_transform/src/bundles.rs:34]:
```suggestion
    note = "Use the `Transform` component instead. Inserting `Transform` will now also insert a `GlobalTransform` automatically."
```
Reworded so people don't need to know what a 'required' component is to understand

File: errors/B0004.md
Comment (tim-blackbird) [inline errors/B0004.md:105]:
```suggestion
However, it will be automatically inserted whenever `Transform` is
```
I think we should also keep the lines you removed above this. It's less likely to be an issue now but it's still true.

**Review by tim-blackbird — COMMENTED** [review]
That's a nice diff c:

Cart *was* talking about removing `SpatialBundle` but as far as I can tell it is still necessary.

File: crates/bevy_transform/src/bundles.rs
Comment (ecoskey) [reply crates/bevy_transform/src/bundles.rs:31]: I'll revert the other bundles as well, meaning the migration will just be updated docs, deprecation, and removing other engine uses of TransformBundle/explicitly adding GlobalTransform.

**Comment by ecoskey** [comment]
> Cart _was_ talking about removing `SpatialBundle` but as far as I can tell it is still necessary.

Still necessary as of this PR, but afaict there's consensus that it'll be replaced by just adding `Transform` and `Visibility` separately

**Comment by ecoskey** [comment]
CI should be fixed, I was a little bit dumb and had `GlobalTransform` require `GlobalTransform` while doing some `git` shenanigans

**Review by TrialDragon — APPROVED** [review]
This LGTM, now that the CI is passing. I'm happy to see the nicer code this is already producing compared to bundles.

File: crates/bevy_transform/src/lib.rs
Comment (ManevilleF) [inline crates/bevy_transform/src/lib.rs:39]: Should we use the new `expect` from rust 1.81.0 ?

Comment (ecoskey) [reply crates/bevy_transform/src/lib.rs:39]: Last I checked the MSRV hadn't been bumped yet. Has it?

**Review by Olle-Lukowski — APPROVED** [review]

File: crates/bevy_transform/src/lib.rs
Comment (alice-i-cecile) [reply crates/bevy_transform/src/lib.rs:39]: Feel free to bump it and use it here. Bevy's policy is "latest stable".

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
This LGTM, but I'm not going to merge until we have consensus from @cart :) Thanks for tackling this!

**Comment by alice-i-cecile** [comment]
@ecoskey once CI is passing, please ping me. We're ready to proceed with this work :) See [this message from Cart on Discord](https://discord.com/channels/691052431525675048/1264881140007702558/1289029221070409738).

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1699 if you'd like to help out.

**Pull request merged**
